File: ai/young_adult.py
# Class worked on by: Claire
import random
from personality import Personality
from data_pre_processor import DataPreProcessor

class YoungAdult(Personality):

    def __init__(self, name):
        Personality.__init__(self, name)
        self.age = random.randint(22, 28)
        self.probSpellingMistake = random.randint(2,5)
    
    def addPersonality(self, input):
        output = self.shortenWords(input)
        return output

    # You -> u for instance, more likely to be commonly appropriate. Needs to return a string. - Kishore
    def shortenWords(self, input):
        shortWords = {
            'for your information' : 'fyi',
            'okay' : 'kk',
            'ok':'kk',
            'easy' : 'ez',
            'see you' : 'cya',
            'because' : 'cuz',
            'very' : 'v',
            'to be honest' : 'tbh',
            'oh my god' : 'OMG',
            'boyfriend' : 'bf',
            'girlfriend' : 'gf',
            'awesome' : 'cool',
            'in trouble' : 'screwed',
            }        
        preProcessor = DataPreProcessor(input) 
        input = preProcessor.string2Array(input)    
        for x in range(0,len(input)):
            for y in shortWords:
                if(str(input[x]) == y):
                    input[x] = shortWords[y]
        input = preProcessor.array2String(input)               
    
        return input
    
    # Add appropriate emjois - Luiz Fellipe
    def addEmojis(self, input):

        if input.find('sad')!= -1:
            index = input.find('sad')
            length = len('sad')
            output_line = input[:index+length] + ' \U0001F625' + input[index+length:]
            print(output_line)
            return output_line
        elif input.find('happy')!= -1:
            index = input.find('happy')
            length = len('happy')
            output_line = input[:index+length] + ' \U0001F603' + input[index+length:]
            print(output_line)
            return output_line
        elif input.find('funny')!= -1:
            index = input.find('funny')
            length = len('funny')
            output_line = input[:index+length] + ' \U0001F602' + input[index+length:]
            print(output_line)
            return output_line
        elif input.find('love')!= -1:
            index = input.find('love')
            length = len('love')
            output_line = input[:index+length] + ' \U0001F60D' + input[index+length:]
            print(output_line)
            return output_line
        elif input.find('star')!= -1:
            index = input.find('star')
            length = len('star')
            output_line = input[:index+length] + ' \U0001F929' + input[index+length:]
            print(output_line)
            return output_line
        elif input.find('good night')!= -1:
            index = input.find('good night')
            length = len('good night')
            output_line = input[:index+length] + ' \U0001F634' + input[index+length:]
            print(output_line)
            return output_line
        elif input.find('angry')!= -1:
            index = input.find('angry')
            length = len('angry')
            output_line = input[:index+length] + ' \U0001F621' + input[index+length:]
            print(output_line)
            return output_line
        
        return input

File: client/TheTuringGameDemo/app/src/main/java/com/sweng/theturinggamedemo/YoungAdult.java
// Class worked on by the AI Group: Claire McNamara

package com.sweng.theturinggamedemo;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Random;
import java.util.regex.Pattern;

public class YoungAdult extends BotPersonality {

    public int age, probSpellingMistake;
    private HashMap<String, char[]> emojis = new HashMap<>();

    public YoungAdult(String name) {
        super(name);
        Random rand = new Random();
        this.age = rand.nextInt((21 - 13) + 1) + 13;
        this.probSpellingMistake = rand.nextInt((5 - 1) + 1) + 5;
        createEmojiHashMap();
    }

    @Override
    public String addPersonality(String input){
        input = shortenWords(input);
        input = addEmpojis(input);
        return input;
    }

    public void createEmojiHashMap(){
        emojis.put("sad", ":(".toCharArray());
        emojis.put("happy", ":)".toCharArray());
        emojis.put("funny", ":')".toCharArray());
        emojis.put("love", "<3".toCharArray());
    }

    // Adds in slang into the AI's message.
    private String shortenWords(String input){
        HashMap<String, String>  shortenedWords = new HashMap<>();
        shortenedWords.put("for your information", "fyi");
        shortenedWords.put("okay", "kk");
        shortenedWords.put("ok", "kk");
        shortenedWords.put("easy", "ez");
        shortenedWords.put("see you", "cya");
        shortenedWords.put("very", "v");
        shortenedWords.put("because", "cuz");
        shortenedWords.put("to be honest", "tbh");
        shortenedWords.put("oh my god", "OMG");
        shortenedWords.put("boyfriend", "bf");
        shortenedWords.put("girlfriend", "gf");
        shortenedWords.put("awesome", "cool");
        shortenedWords.put("in trouble", "screwed");

        String[] inputAr = input.split(" ");
        int wordIndex = 0;
        for (String word:inputAr) {
            if (shortenedWords.containsKey(word)){
                inputAr[wordIndex] = shortenedWords.get(word);
            }
            wordIndex++;
        }
        input = String.join(" ", inputAr);
        return input;
    }

    // Adds emojis where appropriate in a sentence i.e. after the first punctuation character, or at
    // the end of a sentence.
    public String addEmpojis(String input){
        Iterator iter = emojis.entrySet().iterator();
        boolean found = false;
        while (iter.hasNext() && !found){
            Map.Entry elem = (Map.Entry)iter.next();
            String word = (String)elem.getKey();
            if (input.indexOf(word) != -1){
                int index = input.indexOf(word) + word.length();
                input = findPunctuation(input, (char[])elem.getValue(), index);
                found = true;
            }
        }
        return input;
    }

    public String findPunctuation(String input, char[] emoji, int index){
        StringBuilder tmpInput = new StringBuilder(input);
        boolean found = false;
        for (int i = index; i < input.length() && !found; i++){
            if (checkPunctuation(Character.toString(input.charAt(i)))){
                tmpInput.insert(i+1, emoji, 0, 2);
                found = true;
            }
        }
        if (!found) tmpInput.append(emoji);
        input = tmpInput.toString();
        return input;
    }

    public Boolean checkPunctuation(String input){
        return Pattern.matches("[!.,]", input);
    }
}